packages/expo-background-task/android/src/main/java/expo/modules/backgroundtask/BackgroundTaskScheduler.kt

```diff
@@ -0,0 +1,214 @@
+        false
+      }
+    }
+
+    /**
+     Cancels the worker task
+     */
+    suspend fun stopWorker(context: Context): Boolean {
+      Log.i(TAG, "Cancelling worker with identifier $WORKER_IDENTIFIER")
+
+      // Stop our main worker
+      val workManager = WorkManager.getInstance(context)
+      return try {
+        workManager.cancelUniqueWork(WORKER_IDENTIFIER).await()
+        workManager.pruneWork().await()
+      } catch (e: Exception) {
+        Log.i(TAG, "Stopping worker failed with error " + e.message)
+        false
+      }
+    }
+
+    /**
+     Returns true if the worker task is pending
+     */
+    suspend fun isWorkerRunning(context: Context): Boolean {
+      val workInfo = getWorkerInfo(context)
+      return workInfo?.state == WorkInfo.State.RUNNING ||
+        workInfo?.state == WorkInfo.State.ENQUEUED
+    }
+
+    /**
+     * Runs tasks with the given appScopeKey
+     */
+    suspend fun runTasks(context: Context, appScopeKey: String) {
+      // Get task service
+      val taskService = TaskServiceProviderHelper.getTaskServiceImpl(context)
+        ?: throw MissingTaskServiceException()
+
+      Log.i(TAG, "runTasks: $appScopeKey")
+
+      // Get all task consumers
+      val consumers = taskService.getTaskConsumers(appScopeKey)
+      Log.i(TAG, "runTasks: number of consumers ${consumers.size}")
+
+      val tasks = consumers.mapNotNull { consumer ->
+        if (consumer.taskType() == BackgroundTaskConsumer.BACKGROUND_TASK_TYPE) {
+          val bgTaskConsumer = (consumer as? BackgroundTaskConsumer) ?: throw InvalidBackgroundTaskConsumer()
+          Log.i(TAG, "runTasks: executing tasks for consumer of type ${consumer.taskType()}")
+
+          val taskCompletion = CompletableDeferred<Unit>()
+
+          bgTaskConsumer.executeTask {
+            Log.i(TAG, "Task successfully finished")
+            taskCompletion.complete(Unit)
+          }
+
+          taskCompletion
+        } else {
+          null
+        }
+      }
+
+      // Await all tasks to complete
+      tasks.forEach { it.await() }
+    }
+
+    /**
+     * Returns the worker info object from the WorkManager if the worker has been
+     * registered, otherwise returns null
+     */
+    private suspend fun getWorkerInfo(context: Context): WorkInfo? {
+      // Get work manager
+      val workManager = WorkManager.getInstance(context)
+
+      return try {
+        val workInfos = workManager.getWorkInfosForUniqueWork(WORKER_IDENTIFIER).await()
+        return workInfos.firstOrNull()
+      } catch (e: Exception) {
+        Log.i(TAG, "Calling getWorkInfosForUniqueWork failed with error " + e.message)
+        null
+      }
+    }
+
+    /**
+     * Helper function for calling functions returning an Operation
+     */
+    suspend fun Operation.await(): Boolean = suspendCancellableCoroutine { continuation ->
+      val future = this.result
+      val executor = Executors.newCachedThreadPool()
```

Comment:
I'm not sure if creating a thread pool for each operation is a good idea, but I don't know how we can change it.

Reply:
I made the thread pool a companion object on the class - so that it doesn't create a new one each time the function is called.

Nice catch!